Draft policy on the usage of generative AI tools (#177)
fatiando/community: CONTRIBUTING.md

  - [Testing your code](#testing-your-code)
  - [Documentation](#documentation)
  - [Code Review](#code-review)

> * [Policy on Usage of Generative AI Tools](#policy-on-usage-of-generative-ai-tools)

Reviewer: ```suggestion
* [Usage of Generative AI Tools](#usage-of-generative-ai-tools)
```

Try to get them all passing (green).
If you have any trouble, leave a comment in the PR or
[get in touch](#how-can-i-talk-to-you).

> ## Policy on Usage of Generative AI Tools

Reviewer: ```suggestion
## Usage of Generative AI Tools
```

One of the Fatiando a Terra project's visions is to spread awareness about the
value of high quality scientific software, and to build knowledge and train
scientists and developers on the best practices for research software
development.

We value the creation process as a goal itself: by drafting, improving,
redesigning, failing, and fixing we develop that knowledge, we improve our
skills, and we can share it with the rest of the community.
For this reason, **we prefer contributions with human-created content**, as in
code, documentation, guides, figures, videos, etc.

Generative AI tools (genAI tools) are being widely used for generating these
types of content. We understand that some contributors might find them useful,
and would like to include them in their workflows and use them when

> contributing to open-source projects.

Reviewer: Move this paragraph to the start of this section. Then everything else flows a bit better.

Reviewer: Also might be worth 
```suggestion
Generative AI tools (genAI tools, including chatbots and coding agents) are being widely used for generating these
types of content. We understand that some contributors might find them useful,
and would like to include them in their workflows and use them when
contributing to open-source projects.
```

Usage of genAI tools in contributions to Fatiando a Terra are not forbidden,
but when being used as an aid in developing code and editing documentation,

> **contributors must**:

Reviewer: I'd rephrase this to make it clear that the usage is not forbidden exactly but strongly discouraged. 

```suggestion
Usage of genAI tools in contributions to Fatiando a Terra is **strongly discouraged** but are not forbidden (see below).
When such tools are used, **contributors must**:
```

1. ensure that they fully understand the proposed changes and can explain why
   they are correct, and

> 2. communicate that genAI tools have been used in the process.

Reviewer: ```suggestion
1. Ensure that they **fully understand the proposed changes** and can explain why
   they are correct with the aid of a bot.
2. **Declare** that genAI tools have been used in the process (this is covered in our PR template).
```

Some allowed usages of genAI tools are:

> - ✅ Translating text from one language to another.

Reviewer: ```suggestion
- ✅ Translating text from one language to another. The text should be revised to check that the original meaning was not lost.
```

- ✅ Reviewing **your code** to improve it or find potential issues.
- ✅ Exploring new design ideas and different implementations.

We forbid the usage of genAI tools in the following cases:

- ❌ Use of AI Agents (bots) to automatically open or comment on Pull Requests,
  Issues, and any communication channel of the project.
- ❌ Pasting outputs of genAI content directly into Pull Requests, Issues, and

>   any communication channel of the project.

Reviewer: ```suggestion
- ❌ Pasting outputs of genAI content without human revision directly into Pull Requests, Issues, and
  any communication channel of the project.
```

> - ❌ Use of genAI tools to circumvent license restrictions. For example,

Reviewer 2: The other points in this section I think are good, but this seems a bit niche, at least it's not something I've heard about. But if it's a common issue, then all good to leave it!

Author: Yes, I know. I wanted to list it still, just to make it explicit in case someone plans to use AI tools to build a reimplementation of a piece of software, particularly old academic software that original researchers had not released under open-source licenses.

The most relevant case of "license laundry" is what happened in `chardet`: a maintainer overwrote the entire LGPL-licensed code base with an AI generated output and released it under a more permissive license. Leaving a link here to the Issue raising the concern: `https://github.com/chardet/chardet/issues/327`.

Reviewer: I think this is relevant to free software but not really in the context of a project like our right? Unless you think someone could do this as a module in a package?

Reviewer: I'm OK with leaving this in since it's not doing any harm.

Author: The case I was considering (I recognize it can be a bit niche) is for example if someone wants to bring some pre-existing code (e.g. "my advisor wrote this code 15 years ago, would be nice to have it here") into our libraries with either a non open-source license or a license with copyleft. If the original license won't allow it, we cannot take add it. But now that I see people using LLMs to circumvent those licenses by asking the models to rewrite it entirely, I'm worried that someone might be tempted to do so and open a PR with such code.

  reimplementing a GPL-licensed code to release it under a more permissive
  license.
- ❌ Use of genAI tools in code reviews in any of our repositories. We highly
  value the *peer-review* process as a way for everyone to learn from each

>   other's expertise.

Reviewer 2: Maybe we could offer an alternative here, and say something like: As an alternative to genAI tools for code review, we recommend code analysis tools such as PyLint or Ruff, which can catch many common code issues.

Author: Sure! We mention Ruff as part of the normal workflow. We can mention it here again.

Author: Now that I reread this item, I think not allowing _genAI tools in code reviews_ is too broad. What I mean is not to ask a chatbot to review a whole PR. But acceptable uses of genAI during a code review process would be to check for grammar, write snippets that would show how the solution is not covering all cases, helping to write additional tests, etc.

I'm open to suggestions to improve this sentences.

>   other's expertise.

Reviewer: How about:

```suggestion
- ❌ Use of genAI tools to review entire pull requests in any of our repositories. We highly
  value the *peer-review* process as a way for everyone to learn from each
  other's expertise. It would be OK to use the tool to check for grammar, for example, but not write the entire review.
```

Some other uses are not forbidden, but we should pay strong attention.

> For example, using genAI tools to:

Reviewer: I'd include all of these in the forbidden category. Having a PR written by an agent when the person doesn't understand the code is basically item 1 of the list above. Same for documentation. Honestly, I don't want to waste my time reviewing text from an LLM.

Author: I think they do need a rewrite. My intention with these _warning_ points was to cover what I expect to be the majority of cases in which contributors will use LLMs, that could easily slide to unacceptable use cases if not treated with care. For example, asking an LLM to draft a structure for a new class. If contributors will be making such use of LLMs, it's impossible for us to detect whether that's the case or not. But as long as that code is well understood and reviewed by the author (basically, using the LLM as a _glorified autocomplete_), I think that use case should be allowed.

I think we should move the idea of "LLMs might save you some typing time, but we'll need to spend the time reviewing the submission. Typing type is not the bottleneck in our projects" to one of the first paragraphs. To make it clear that use of LLMs usually shift the workload from the contributor to the maintainer.

What do you think?

Reviewer: Yes, I agree. Making it clear at the start is good. I'd add that we are not in any rush and our approach is to think things through very well before implementation. So people sending us PRs with no thought doesn't help at all.

I would still move these warning points to the forbidden category. When you wrote these, I imagine you thought of people who are thoughtful and parts of our community using AI to speedup their process. But that's not what we've been seeing. What we're getting is people who know nothing about the problem domain submitting dozens of LLM-generated PRs a day to several projects to boost their profiles.

- ⚠️ Write code for us. While they can save some *typing time*, one needs to
  spend time reviewing the generated output and making sure we fully understand

>   the proposed solution.

Reviewer: ```suggestion
- ⚠️ Write code for us. While they can save you some *typing time*, we need to
  spend time reviewing the generated output to make sure we fully understand
  the proposed solution and then maintain it in the future. Typing time is not the bottleneck in our projects.
```

- ⚠️ Write code for us that we couldn't write by ourselves. This means that
  it's possible that we don't fully understand the generated code.
- ⚠️ Write documentation for us. Documenting code can help identifying design
  issues and non-intended uses of the code. Make sure the generated text is
  accurate, and describes well how the piece of code should be used.
- ⚠️ Ask for references on papers or books. Double check those references do
  exists, and make sure they are relevant for the matter.

Maintainers reserve the right, at their sole discretion, to close Pull
Requests, Issues, and delete comments that incur in unacceptable usage of genAI

> tools.

Reviewer: ```suggestion
The lists above are not exhaustive and serve as examples.
Maintainers reserve the right, at their sole discretion, to close Pull
Requests, Issues, and delete comments that they judge to incur in unacceptable usage of genAI
tools.
```
